docs/developer/architecture/centralized-cli.rst

```diff
@@ -149,3 +149,203 @@ Cluster Service Administration
+Package for Operating System supported by MetalK8s.
+
+The package should install the plugin in ``/usr/bin`` directory by default.
+
+This plugin should be installed on the bootstrap node by default.
+
+Rejected design choice
+""""""""""""""""""""""
+
+- ``Bash`` kubectl plugin: Bash is great to do simple actions but not when
+  you need to do interaction with some API like Kubernetes API or Salt API.
+- ``Python`` kubectl plugin: Python allows us to do complicated actions and
+  great interactions with APIs but interactions between ``Go`` and Kubernetes
+  are much better, with a sample CLI available.
+
+``metalk8sctl`` CLI
+^^^^^^^^^^^^^^^^^^^
+
+Language
+""""""""
+
+``Python`` is chosen as the language for kubectl plugin for the following
+reasons:
+
+- Ability to interact with Salt directly with Python API.
+- Python installation needed anyway by Salt-minion.
+
+Input and Output
+""""""""""""""""
+
+- Command style:
+
+  .. code-block:: bash
+
+    metalk8sctl <resource> <action>
+
+- Interactive:
+
+  Never.
+
+- Output style:
+
+  Human readable output, do not necessarily need for "machine output" like
+  JSON and YAML.
+
+  Output should extract useful information from Salt output when needed, in
+  case of error only output useful error message(s) from Salt output
+
+.. todo::
+
+  - Define logging information
+
+Salt interaction
+""""""""""""""""
+
+All Salt interaction should be done using
+`Salt Python API <https://docs.saltstack.com/en/latest/ref/clients/index.html>`_
+and not use the ``salt-call``, ``salt``, ``salt-run`` binary at all.
+
+This Python API allows us to interact with Salt-master directly from the host
+machine as Python API directly actis on the Salt socket and does not need to
+execute a command inside the Salt-master container.
+
+Deployment
+""""""""""
+
+``metalk8sctl`` should be available from the ISO and also as a System
+Package for Operating System supported by MetalK8s.
+
+This package should be installed on the bootstrap node by default.
+
+Rejected design choice
+""""""""""""""""""""""
+
+- ``bash`` MetalK8s CLI: Bash is great to do simple actions but not when you
+  need to do interaction with Salt, Salt API, and Kubernetes API.
+- Do not follow kubectl style for the command (``<action> <resource>``), it
+  does not make sense to regroup command per action as actions are really
+  different and this CLI only manages a few resources.
+
+Implementation Details
+----------------------
+
+Two different projects.
+
+First have a simple framework to implement a simple command,
+then each command would update the framework if needed.
+
+TBD !
+
+Iteration 1
+^^^^^^^^^^^
+
+``metalk8sctl``: (everything ?)
+
+- bootstrap deploy
+- bootstrap import-iso ?
+- cluster upgrade ?
+- cluster downgrade ?
+- etcd health
+- bootstrap backup ?
+- bootstrap restore ?
+- solutions import-iso ?
+
+Iteration 2
+^^^^^^^^^^^
+
+``metalk8s`` kubectl plugin: (everything ?)
```

Comment:
We don't have `get` or `describe` commands? Could be useful to show Salt-related info (like grains/pillar for a Node's minion), or adjust how some things are displayed.

I'd love to get some wrappers on Environments, which are just groups of namespaces with a common label:

```shell
kubectl metalk8s get env
NAME    NAMESPACES                  SOLUTIONS
prod    prod-storage, prod-zenko    storage(1.0.2), zenko(2.0.1)
test    test                        storage(1.0.2), zenko(2.1.0-rc1)

kubectl metalk8s get env prod -o json
{
  "name": "prod",
  "namespaces": ["prod-storage", "prod-zenko"],
  "solutions": [
    {
      "name": "storage",
      "version": "1.0.2",
      "operator": "prod-storage/storage-operator",
      "ui": "prod-storage/storage-admin-ui",
    },
    ...
  ]
}

# We could even provide wrappers to the usual get/delete/... commands
kubectl metalk8s get deploy --env prod
#   - This would return Deployments in both prod-storage and prod-zenko
```

Reply:
Yes true we should have some get, describe, etc. I just took command from the requirements for the moment.

Not sure we want to list all "utility commands" here ?

Reply:
Let's list all ideas in some table, and only implement a subset in the first iteration. But yeah, having all the "possible" commands can help us adjust the design for the first commands we'll implement.

Reply:
Done in requirement, I have updated the list of commands